File: api/compute/contours.py
from http.server import BaseHTTPRequestHandler
import json
import math

class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        length = int(self.headers['Content-Length'])
        body = json.loads(self.rfile.read(length))

        points = body.get('points', [])
        interval = float(body.get('interval', 1.0))

        if len(points) < 3:
            self.send_response(400)
            self.send_header('Content-type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({
                'data': None,
                'error': 'Minimum 3 points required for contour generation'
            }).encode())
            return

        try:
            pts = [(p['easting'], p['northing'], float(p.get('rl', 0))) for p in points]
            elevations = [p[2] for p in pts]

            tri = delaunay_triangulation(pts)

            z_min = min(elevations)
            z_max = max(elevations)
            levels = []
            start = math.ceil(z_min / interval) * interval
            while start < z_max:
                levels.append(start)
                start += interval

            contours = []
            for level in levels:
                segments = []
                for simplex in tri:
                    v0, v1, v2 = simplex
                    crossings = []

                    edges = [(v0, v1), (v1, v2), (v2, v0)]
                    for va, vb in edges:
                        za, zb = va[2], vb[2]
                        if (za <= level < zb) or (zb <= level < za):
                            t = (level - za) / (zb - za)
                            x = va[0] + t * (vb[0] - va[0])
                            y = va[1] + t * (vb[1] - va[1])
                            crossings.append([round(x, 3), round(y, 3)])

                    if len(crossings) == 2:
                        segments.append(crossings)

                if segments:
                    major = abs(level % (interval * 5)) < 0.001
                    contours.append({
                        'level': round(level, 3),
                        'segments': segments,
                        'major': major
                    })

            self.send_response(200)
            self.send_header('Content-type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({
                'data': {
                    'contours': contours,
                    'bounds': {
                        'z_min': z_min,
                        'z_max': z_max,
                        'interval': interval
                    }
                },
                'error': None
            }).encode())

        except Exception as e:
            self.send_response(500)
            self.send_header('Content-type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({
                'data': None,
                'error': f'An error occurred during contour generation: {str(e)}',
                'meta': { 'fallback': True }
            }).encode())


def delaunay_triangulation(points):
    # Bowyer-Watson algorithm — O(n^2), pure Python implementation
    # Source: Bowyer, A. (1981). "Computing Dirichlet tessellations."
    #         Watson, D.F. (1981). "Computing the n-dimensional Delaunay tessellation."
    n = len(points)
    if n < 3:
        return []

    # Build super-triangle that contains all points
    min_x = min(p[0] for p in points)
    max_x = max(p[0] for p in points)
    min_y = min(p[1] for p in points)
    max_y = max(p[1] for p in points)
    dx = max_x - min_x or 1.0
    dy = max_y - min_y or 1.0
    delta_max = max(dx, dy) * 2

    mid_x = (min_x + max_x) / 2
    mid_y = (min_y + max_y) / 2

    p0 = (mid_x - delta_max, mid_y - delta_max, 0)
    p1 = (mid_x, mid_y + delta_max, 0)
    p2 = (mid_x + delta_max, mid_y - delta_max, 0)

    triangles = [(p0, p1, p2)]

    for point in points:
        bad_triangles = []
        for tri in triangles:
            if is_point_in_circumcircle(point, tri):
                bad_triangles.append(tri)

        boundary = []
        for tri in bad_triangles:
            for edge in [(tri[0], tri[1]), (tri[1], tri[2]), (tri[2], tri[0])]:
                if not any(is_same_edge(edge, e) for e in boundary):
                    boundary.append(edge)

        triangles = [t for t in triangles if t not in bad_triangles]

        for edge in boundary:
            triangles.append((edge[0], edge[1], point))

    # Remove triangles containing super-triangle vertices
    result = []
    for tri in triangles:
        if not any(
            v[0] == p0[0] or v[0] == p1[0] or v[0] == p2[0]
            for v in tri
        ):
            result.append(tri)

    return result


def is_point_in_circumcircle(point, triangle):
    ax, ay, _ = triangle[0]
    bx, by, _ = triangle[1]
    cx, cy, _ = triangle[2]
    px, py, _ = point

    det = (
        (ax**2 + ay**2) * (by - cy) +
        (bx**2 + by**2) * (cy - ay) +
        (cx**2 + cy**2) * (ay - by)
    )
    if abs(det) < 1e-12:
        return False

    sign = (
        (ax**2 + ay**2) * (by - cy) +
        (bx**2 + by**2) * (cy - ay) +
        (cx**2 + cy**2) * (ay - by)
    )

    return sign * det > 0


def is_same_edge(e1, e2):
    return (e1[0] == e2[0] and e1[1] == e2[1]) or \
           (e1[0] == e2[1] and e1[1] == e2[0])

File: src/lib/engine/contours.ts
export interface SpotHeight {
  name: string
  easting: number
  northing: number
  elevation: number
}

export interface ContourLine {
  elevation: number
  points: Array<{ easting: number; northing: number }>
}

export interface Triangle {
  p1: SpotHeight
  p2: SpotHeight
  p3: SpotHeight
}

export function triangulate(points: SpotHeight[]): Triangle[] {
  const triangles: Triangle[] = []

  if (points.length < 3) return triangles

  for (let i = 0; i < points.length - 2; i++) {
    for (let j = i + 1; j < points.length - 1; j++) {
      for (let k = j + 1; k < points.length; k++) {
        const p1 = points[i]
        const p2 = points[j]
        const p3 = points[k]

        const d12 = Math.sqrt((p2.easting-p1.easting)**2 + (p2.northing-p1.northing)**2)
        const d23 = Math.sqrt((p3.easting-p2.easting)**2 + (p3.northing-p2.northing)**2)
        const d31 = Math.sqrt((p1.easting-p3.easting)**2 + (p1.northing-p3.northing)**2)
        const perimeter = d12 + d23 + d31
        const maxSide = Math.max(d12, d23, d31)

        if (maxSide > perimeter * 0.7) continue

        triangles.push({ p1, p2, p3 })
      }
    }
  }

  return triangles
}

function interpolateEdge(
  p1: SpotHeight,
  p2: SpotHeight,
  contourElev: number
): { easting: number; northing: number } | null {
  const minElev = Math.min(p1.elevation, p2.elevation)
  const maxElev = Math.max(p1.elevation, p2.elevation)

  if (contourElev <= minElev || contourElev >= maxElev) return null

  const t = (contourElev - p1.elevation) / (p2.elevation - p1.elevation)

  return {
    easting: p1.easting + t * (p2.easting - p1.easting),
    northing: p1.northing + t * (p2.northing - p1.northing)
  }
}

export function generateContours(
  points: SpotHeight[],
  interval: number
): ContourLine[] {
  if (points.length < 3) return []

  const elevations = points.map((p: any) => p.elevation)
  const minElev = Math.min(...elevations)
  const maxElev = Math.max(...elevations)

  const firstContour = Math.ceil(minElev / interval) * interval
  const contourElevations: number[] = []
  for (let e = firstContour; e <= maxElev; e += interval) {
    contourElevations.push(e)
  }

  const triangles = triangulate(points)
  const contourLines: ContourLine[] = []

  for (const contourElev of contourElevations) {
    const contourPoints: Array<{ easting: number; northing: number }> = []

    for (const tri of triangles) {
      const edges: Array<[SpotHeight, SpotHeight]> = [
        [tri.p1, tri.p2],
        [tri.p2, tri.p3],
        [tri.p3, tri.p1]
      ]

      const crossings: Array<{ easting: number; northing: number }> = []

      for (const [a, b] of edges) {
        const crossing = interpolateEdge(a, b, contourElev)
        if (crossing) crossings.push(crossing)
      }

      if (crossings.length === 2) {
        contourPoints.push(crossings[0], crossings[1])
      }
    }

    if (contourPoints.length > 0) {
      contourLines.push({
        elevation: contourElev,
        points: contourPoints
      })
    }
  }

  return contourLines
}
